jenkins-plugin/src/test/java/org/jenkinsci/plugins/pipeline/maven/WithMavenStepTest.java

```diff
@@ -29,79 +29,164 @@
-                WorkflowRun b = p.scheduleBuild2(0).waitForStart();
-                SemaphoreStep.waitForStart("wait/1", b);
-            }
-        });
-        rr.addStep(new Statement() {
-            @Override
-            public void evaluate() throws Throwable {
-                WorkflowJob p = rr.j.jenkins.getItemByFullName("p", WorkflowJob.class);
-                WorkflowRun b = p.getBuildByNumber(1);
-                SemaphoreStep.success("wait/1", null);
-                rr.j.assertBuildStatusSuccess(rr.j.waitForCompletion(b));
-                SemaphoreStep.success("wait/2", null);
-                rr.j.buildAndAssertSuccess(p);
-            }
-        });
     }
 
     @Issue("SECURITY-441")
     @Test
-    public void fileOnMaster() {
-        rr.addStep(new Statement() {
-            @Override
-            public void evaluate() throws Throwable {
-                JavaContainer slaveContainer = slaveRule.get();
-                rr.j.jenkins.addNode(new DumbSlave("remote", "", "/home/test/slave", "1", Node.Mode.NORMAL, "",
-                                                   new SSHLauncher(slaveContainer.ipBound(22), slaveContainer.port(22), "test", "test", "", ""),
-                                                   RetentionStrategy.INSTANCE, Collections.<NodeProperty<?>>emptyList()));
-                File onMaster = new File(rr.j.jenkins.getRootDir(), "onmaster");
-                String secret = "secret content on master";
-                FileUtils.writeStringToFile(onMaster, secret);
-                WorkflowJob p = rr.j.createProject(WorkflowJob.class, "p");
-                String pipelineScript = "node('remote') {withMaven(mavenSettingsFilePath: '" + onMaster + "') {echo readFile(MVN_SETTINGS)}}";
-                p.setDefinition(new CpsFlowDefinition(pipelineScript, true));
-                WorkflowRun b = rr.j.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0));
-                rr.j.assertLogNotContains(secret, b);
-            }
-        });
+    public void maven_build_on_remote_agent_with_settings_file_on_master_fails() throws Exception {
+        File onMaster = new File(jenkinsRule.jenkins.getRootDir(), "onmaster");
+        String secret = "secret content on master";
+        FileUtils.writeStringToFile(onMaster, secret);
+        String pipelineScript = "node('remote') {withMaven(mavenSettingsFilePath: '" + onMaster + "') {echo readFile(MVN_SETTINGS)}}";
+
+        WorkflowJob p = jenkinsRule.createProject(WorkflowJob.class, "p");
+        p.setDefinition(new CpsFlowDefinition(pipelineScript, true));
+        WorkflowRun b = jenkinsRule.assertBuildStatus(Result.FAILURE, p.scheduleBuild2(0));
+        jenkinsRule.assertLogNotContains(secret, b);
     }
 
+    @Test
+    public void maven_build_on_master_with_specified_maven_installation_succeeds() throws Exception {
+        loadMavenJarProjectInGitRepo(this.gitRepoRule);
+
+        String pipelineScript = "node('master') {\n" +
```

Comment:
Run on the agent by transferring the repository using `git bundle` and `FilePath` copying.